runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/GrpcWindmillServer.java

```diff
@@ -936,7 +936,11 @@ protected void onResponse(StreamingGetWorkResponseChunk chunk) {
               .execute(
                   () -> {
                     try {
-                      send(extension);
+                      synchronized (this) {
+                        if (!clientClosed.get()) {
```

Comment:
Should send itself be changed to check this?

Most sends are wrapped with
 catch (IllegalStateException) 
but then ignore the exception or just log.

Was this causing issues other than unnecessary IllegalStateExceptions?

Reply:
> Was this causing issues other than unnecessary IllegalStateExceptions?

I am investigating stalled GetWork streams with logs like `Output channel stalled for 31s, outbound thread`. I think sending messages on a closed stream has something to do with it. In test runs with this change, I'm yet to see a stalled GetWorkStream.

> Should send itself be changed to check this?

There are a few sends without the catch(IllegalStateException), they need to be looked at individually. Can we get this in first?

Reply:
To keep contract the same for other call-sites you could throw IllegalStateException yourself if the client is closed instead of relying on grpc to do so (since that adds delay it sounds like).

LGTM though if you want to put a comment/TODO for that to try to get this in the next release. You'll have to find someone else to merge though.

Reply:
+1 to throwing `IllegalStateException` if the stream is closed. Otherwise we risk having regressions by changing the behavior. It's safer to throw the exception for the purposes of yoru change.

Reply:
Moved the closed check inside send.